libs/test-utils/src/mock_function.ts

```diff
@@ -167,22 +249,54 @@ export function mockFunction<Func extends AnyFunc>(
   };
 
   mock.assertComplete = () => {
-    if (
-      !deepEqual(
-        state.actualCalls,
-        state.expectedCalls.map(({ input }) => ({ input }))
-      )
-    ) {
-      const numCalls = Math.max(
-        state.actualCalls.length,
-        state.expectedCalls.length
-      );
-      const message = `Mismatch between expected mock function calls and actual mock function calls:\n\n${[
-        ...Array.from({ length: numCalls }).keys(),
-      ]
-        .map((index) => {
-          const expectedCall = state.expectedCalls[index];
-          const actualCall = state.actualCalls[index];
+    // We want to check:
+    // - Every actual call matched an expected call
+    // - Every expected call was used
+    // So we build a correspondence of actual calls with their expected calls.
+
+    const actualCallsWithTheirExpectedCalls = state.actualCalls.map(
+      (actualCall, actualCallIndex) => {
+        const expectedCall =
+          findExpectedCallForActualCallIndex(actualCallIndex);
+        return {
+          actualCall,
+          expectedCall,
+        };
+      }
+    );
+
+    const unusedExpectedCalls = state.expectedCalls.filter(
+      (expectedCall) =>
+        !actualCallsWithTheirExpectedCalls.some(
+          ({ expectedCall: usedExpectedCall }) =>
+            usedExpectedCall === expectedCall
+        )
+    );
+
+    const callCorrespondence = [
+      ...actualCallsWithTheirExpectedCalls,
+      ...unusedExpectedCalls.map((expectedCall) => ({
+        actualCall: undefined,
+        expectedCall,
+      })),
+    ];
+
+    // We can only accurately report a correspondence of expected and actual
+    // calls up to the first mismatched call. After that, we don't know which
+    // expected calls were used for which actual calls.
```

Comment:
This is the only major change to existing functionality. We now no longer show the calls that occurred after a mismatched call. Before, since we expected a 1-1 mapping of actual calls to expected calls, it was easy to just zip the two lists together. Now that there are repeated calls, it's basically impossible to show what was "supposed" to happen after a mismatch.

Reply:
Makes sense! Seeing the calls after the mismatch doesn't provide that much value anyway in my opinion